src/Features/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs

```diff
@@ -0,0 +1,572 @@
+    /// appropriate diagnostics.
+    /// </para>
+    /// </summary>
+    /// <remarks>
+    /// Note: the json structure we parse out is actually very simple.  It's effectively all lists
+    /// of <see cref="JsonValueNode"/> values.  We just treat almost everything as a 'value'.  For
+    /// example, a <see cref="JsonPropertyNode"/> (i.e. <c>"x" = 0</c>) is a 'value'.  As such, it
+    /// can show up in arrays (i.e.  <c>["x" = 0, "y" = 1]</c>).  This is not legal, but it greatly
+    /// simplifies parsing.  Effectively, we just have recursive list parsing, where we accept any
+    /// sort of value in any sort of context.  A later pass will then report errors for the wrong
+    /// sorts of values showing up in incorrect contexts.
+    /// <para>
+    /// Note: We also treat commas (<c>,</c>) as being a 'value' on its own.  This simplifies parsing
+    /// by allowing us to not have to represent Lists and SeparatedLists.  It also helps model
+    /// things that are supported in json.net (like <c>[1,,2]</c>).  Our post-parsing pass will
+    /// then ensure that these comma-values only show up in the right contexts.
+    /// </para>
+    /// </remarks>
+    [NonCopyable]
+    internal partial struct JsonParser
+    {
+        private static readonly string s_closeBracketExpected = string.Format(FeaturesResources._0_expected, ']');
+        private static readonly string s_closeBraceExpected = string.Format(FeaturesResources._0_expected, '}');
+        private static readonly string s_openParenExpected = string.Format(FeaturesResources._0_expected, '(');
+        private static readonly string s_closeParenExpected = string.Format(FeaturesResources._0_expected, ')');
+        private static readonly string s_commaExpected = string.Format(FeaturesResources._0_expected, ',');
+
+        private JsonLexer _lexer;
+        private JsonToken _currentToken;
+        private int _recursionDepth;
+
+        // Fields used to keep track of what types of json values we're in.  They're used for error
+        // recovery, specifically with respect to encountering unexpected tokens while parsing out a
+        // sequence of values.  For example, if we have:  ```{ a: [1, 2, }```, we will mark that
+        // we're both in an object and in an array.  When we then encounter the errant ```}```,
+        // we'll see that we were in an object, and thus should stop parsing out the sequence for
+        // the array so that the ```}``` can be consume by the object we were in.  However, if we
+        // just had ```[1, 2, }```, we would not be in an object, and we would just consume the
+        // ```}``` as a bogus value inside the array.
+        //
+        // This approach of keeping track of the parse contexts we're in, and using them to
+        // determine if we should consume or pop-out when encountering an error token, mirrors the
+        // same approach that we use in the C# and TS/JS parsers.
+        private bool _inObject;
```

Comment:
since I'm using a browser I'm just going to pose this as question since it's hard to verify: 

Are there tests for embedded sequences to make sure we still get the behavior we want? `[ 1, [ 2, [ 3]] ]` and other weird things like `[ 1, { 2, [ 3, { 4 }]}]` and various permutations of the closing tokens

Reply:
there are def tests for the former.  but i'll beef both sorts up now.